dali/operators/nvcvop/nvcvop.h

```diff
@@ -0,0 +1,227 @@
+class NVCVOperator: public BaseOp {
+ public:
+  using BaseOp::spec_;
+  using BaseOp::BaseOp;
+
+  /**
+   * @brief Get an operator argument as an nvcv Tensor.
+   * This method uses the dynamic scratchpad to allocate the (pinned and device) buffers
+   * used by the output tensor.
+   *
+   * The scratchpad should return its memory in an order that is compatible with ws.stream().
+   * The output is valid only for the lifetime of the scratchpad.
+   *
+   * @tparam T data type of the DALI operator argument
+   * @tparam DTYPE expected nvcv Tensor data type
+   * @param arg_shape shape of the DALI operator argument.
+   */
+  template <typename T, NVCVDataType DTYPE>
+  nvcv::Tensor AcquireTensorArgument(Workspace &ws, kernels::DynamicScratchpad &scratchpad,
+                                     ArgValue<T, 1> &arg, const TensorShape<> &arg_shape,
+                                     TensorLayout layout = "") {
+    auto dtype = nvcv::DataType(DTYPE);
+    int num_samples = ws.GetInputBatchSize(0);
+    int arg_sample_vol = volume(arg_shape);
+    arg.Acquire(spec_, ws, num_samples, arg_shape);
+    T *staging_buffer =
+        scratchpad.AllocatePinned<T>(num_samples * arg_sample_vol, dtype.alignment());
+    for (int s = 0; s < num_samples; ++s) {
+      std::copy(arg[s].data, arg[s].data + arg_sample_vol,
+                staging_buffer + s * arg_sample_vol);
+    }
+    T *device_buffer = scratchpad.AllocateGPU<T>(num_samples * arg_sample_vol, dtype.alignment());
+    MemCopy(device_buffer, staging_buffer, num_samples * arg_sample_vol * sizeof(T), ws.stream());
+
+    auto shape_data = CalcTensorShape(arg_shape, num_samples, dtype);
+    nvcv::TensorDataStridedCuda::Buffer inBuf;
+    inBuf.basePtr = reinterpret_cast<NVCVByte*>(device_buffer);
+    inBuf.strides[shape_data.size() - 1] = dtype.strideBytes();
+    for (int d = shape_data.size() - 2; d >= 0; --d) {
+      inBuf.strides[d] = shape_data[d + 1] * inBuf.strides[d + 1];
+    }
+    TensorLayout out_layout = layout.empty() ? layout : "N" + layout;
+    nvcv::TensorShape shape(shape_data.data(), shape_data.size(),
+                            nvcv::TensorLayout(out_layout.c_str()));
+    nvcv::TensorDataStridedCuda inData(shape, dtype, inBuf);
+    return nvcv::TensorWrapData(inData);
+  }
+
+  std::vector<int64_t> CalcTensorShape(const TensorShape<> &arg_shape, int num_samples,
+                                       const nvcv::DataType &dtype) {
+    int ndim = arg_shape.sample_dim();
+    int64_t inner_dim = arg_shape[ndim - 1];
+    DALI_ENFORCE(inner_dim % dtype.numChannels() == 0, "Invalid argument shape.");
+    std::vector<int64_t> shape_data;
+    // If number of channels in the data type matches the innermost dimension,
+    // we remove this dimension
+    if (inner_dim == dtype.numChannels()) {
+      shape_data.resize(ndim);
+      shape_data[0] = num_samples;
+      for (int d = 0; d < ndim - 1; ++d) {
+        shape_data[d + 1] = arg_shape[d];
+      }
```

Comment:
I think this logic can lead to all sorts of problems - are we sure that ambiguous cases (like numChannels == 1 && inner_dim == 1) are handled properly?

Reply:
I changed the logic here to be less ambiguous. 
Now there are two cases: dtype has single channel or dtype has multiple channels.
In single-channel case, the resulting shape is equal to the DALI shape.
In multi-channel case, the inner dim of DALI shape **has to be equal** to the number of channels and it's removed.

Generally these two cases come from the fact that cvcuda is not consistent in how it expects the args tensors to be shaped, e.g. window size for median blur is expected to be a {Nx2} tensor of ints and windows size for morphology is expected to be a {N} tensor of int2s.